src/main/java/pl/edu/agh/firecell/engine/algorithm/TemperaturePropagator.java

```diff
@@ -0,0 +1,191 @@
+
+        double newTemperature = currentTemperature;
+
+        switch (oldCell.material()) {
+            case WOOD -> newTemperature = newWoodTemperature(oldCell, currentTemperature, burningTime, newTemperature);
+            case AIR -> newTemperature = newAirTemperature(oldState, oldCell, cellIndex, currentTemperature, newTemperature);
+        }
+        return newTemperature;
+    }
+
+    private double newAirTemperature(State oldState, Cell oldCell, Vector3i cellIndex, double currentTemperature, double newTemperature) {
+        if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).burningTime() > 0 &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).flammable() &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).remainingFirePillar() - 1 > 0) {
+            newTemperature = Math.max(currentTemperature, AIR.getBurningTemperature());
+        }
+
+        if (!oldState.hasCell(NeighbourUtils.down(cellIndex)) ||
+                oldState.getCell(NeighbourUtils.down(cellIndex)).burningTime() == 0 ||
+                !oldState.getCell(NeighbourUtils.down(cellIndex)).flammable()) {
+
+            if (doesNeighbourOnFire(oldState, cellIndex)
+                    .anyMatch(neighbourIndex -> !possibleToGoUp(oldState, neighbourIndex) &&
+                            oldState.getCell(neighbourIndex).remainingFirePillar() - 1 > 0))
+            {
+                newTemperature = Math.max(newTemperature, 600);
+            }
+        }
+
+        // keep air on fire
+        if (oldCell.burningTime() > 0) {
+            newTemperature = Math.max(currentTemperature, 600);
+        }
+        return newTemperature;
+    }
+
+    private double newWoodTemperature(Cell oldCell, double currentTemperature, int burningTime, double newTemperature) {
+        if (currentTemperature > WOOD.autoIgnitionTemperature() &&
+                oldCell.burningTime() == 0 &&
+                MAX_BURNING_TIME != 0 &&
+                oldCell.flammable()) {
+            //set wood on fire
+            newTemperature = WOOD.getBurningTemperature();
+        } else if (burningTime > 0 && oldCell.burningTime() <= MAX_BURNING_TIME) {
+            //keep wood on fire
+            newTemperature = Math.max(WOOD.getBurningTemperature(), currentTemperature);
+        }
+        return newTemperature;
+    }
+
+    private double computeConduction(State oldState, Cell oldCell, Vector3i cellIndex) {
+        return calculateAxisDifference(oldState, NeighbourUtils.north(cellIndex), oldCell, NeighbourUtils.south(cellIndex)) +
+                calculateAxisDifference(oldState, NeighbourUtils.up(cellIndex), oldCell, NeighbourUtils.down(cellIndex)) +
+                calculateAxisDifference(oldState, NeighbourUtils.east(cellIndex), oldCell, NeighbourUtils.west(cellIndex));
```

Comment:
Use axis neighbour streams

Reply:
I do not know how to collect data from stream in due to pass as two arguments to calculateAxisDifference().